### ISSUE TITLE
Adding comments to unclear code

### ISSUE THREAD

**Issue by immerSIR** [223MapAction/Model_deploy#1]
**Description:**
We need to improve the readability of our `predict` function in `app/services/cnn/cnn.py`. This function is crucial for performing multi-label image classification using our pre-loaded ResNet50 model. Adding clear, explanatory comments will help new contributors understand the function's purpose and logic.

This is a great first issue because it allows you to dive into our machine learning codebase, understand a key function, and improve its documentation without changing its functionality.

**Steps to complete:**
1. Locate the `predict` function in `app/services/cnn/cnn.py`
2. Read through the function to understand its purpose and logic
3. Add comments explaining:
   - The purpose of the function
   - What each parameter represents
   - The logic of each step within the function
   - What the function returns and why
4. Ensure your comments are clear, concise, and follow our commenting style (see our `CONTRIBUTING.md` for guidelines)
5. Submit a pull request with your changes

Here's the function you'll be commenting:

```python
def predict(image):
    """
    Performs multi-label image classification using a pre-loaded ResNet50 model.
    
    Args:
        image (bytes): The image data in bytes format.
    
    Returns:
        tuple: A tuple containing a list of predicted tags and a list of probabilities.
    """
    model.eval()  # Set model to evaluation mode
    input_data = preprocess_image(image)  # Preprocess the image
    with torch.no_grad():  # Disable gradient computation
        output = model(input_data)  # Forward pass
    probabilities = torch.sigmoid(output[0])  # Apply sigmoid to get probabilities

    top_predictions = [
        (tags[i], prob.item()) 
        for i, prob in enumerate(probabilities) 
        if prob.item() > 0.4
    ]
    top_predictions.sort(key=lambda x: x[1], reverse=True)
    top_predictions = top_predictions[:3]

    return top_predictions, probabilities.tolist()
```

Add comments explaining what each step of the function does, why we're using `torch.no_grad()`, what the `probabilities` represent, and how we're selecting and sorting the top predictions.

**Example of Enhanced Function with Explanatory Comments:**

```python
def predict(image):
    """
    Performs multi-label image classification using a pre-loaded ResNet50 model.
    
    Args:
        image (bytes): The image data in bytes format.
    
    Returns:
        tuple: A tuple containing a list of predicted tags and a list of probabilities.
    """
    model.eval()  # Set the model to evaluation mode to disable dropout and batch normalization layers
    
    input_data = preprocess_image(image)  # Preprocess the image to match the model's expected input format
    
    with torch.no_grad():  # Disable gradient computation to reduce memory usage and increase inference speed
        output = model(input_data)  # Perform a forward pass through the model to get raw predictions
    
    probabilities = torch.sigmoid(output[0])  # Apply the sigmoid function to convert raw outputs to probabilities between 0 and 1
    
    # Extract tags and their corresponding probabilities that exceed the threshold of 0.4
    top_predictions = [
        (tags[i], prob.item()) 
        for i, prob in enumerate(probabilities) 
        if prob.item() > 0.4
    ]
    
    # Sort the predictions in descending order based on probability
    top_predictions.sort(key=lambda x: x[1], reverse=True)
    
    # Select the top 3 predictions for concise results
    top_predictions = top_predictions[:3]
    
    return top_predictions, probabilities.tolist()  # Return the top predictions and all probabilities as a list
```

**Additional Notes:**
- **Understanding `torch.no_grad()`:** Using `torch.no_grad()` is essential during inference as it disables gradient calculations, which are unnecessary for prediction and help in reducing memory consumption and speeding up the computation.
  
- **Probabilities Interpretation:** The `probabilities` represent the likelihood of each tag being associated with the input image. A value closer to 1 indicates a higher confidence that the tag applies to the image.

- **Selecting Top Predictions:** The function filters out predictions with probabilities below 0.4 to focus on more confident predictions. It then sorts the remaining predictions in descending order of probability and selects the top three to provide the most relevant tags.

**Best Practices:**
- **Clear and Concise Comments:** Ensure that comments are easy to understand and provide meaningful insights into what each part of the code is doing.
  
- **Consistent Commenting Style:** Follow the project's existing commenting conventions as outlined in the `CONTRIBUTING.md` to maintain consistency across the codebase.

- **Avoid Redundant Comments:** While it's important to explain complex logic, avoid over-commenting simple or self-explanatory lines to keep the code clean and readable.

If you need any help or have questions about our coding standards or commenting practices, feel free to ask in the comments!

**Comment by Itz-Agasta** [comment]
Hi @immerSIR ,
I just updated the `cnn.py` file. Could you please review my [pull request](https://github.com/223MapAction/Model_deploy/pull/3) when you have some time?
